filters/auth/tokenintrospection.go

```diff
@@ -0,0 +1,342 @@
+// IntrospectionEndpoint to the given oauthIntrospectionURL.
+func NewOAuthTokenintrospectionAnyKV(cfg *OpenIDConfig, timeout time.Duration) filters.Spec {
+	return newOAuthTokenintrospectionFilter(checkOAuthTokenintrospectionAnyKV, cfg, timeout)
+}
+
+// NewOAuthTokenintrospectionAllKV creates a new auth filter specification
+// to validate authorization for requests. Current implementation uses
+// Bearer tokens to authorize requests and checks that the token
+// contains at least one key value pair provided.
+//
+// This is implementing RFC 7662 compliant implementation. It uses
+// POST requests to call introspection_endpoint to get the information
+// of the token validity.
+//
+// It uses /.well-known/openid-configuration path to the passed
+// oauthIssuerURL to find introspection_endpoint as defined in draft
+// https://tools.ietf.org/html/draft-ietf-oauth-discovery-06, if
+// oauthIntrospectionURL is a non empty string, it will set
+// IntrospectionEndpoint to the given oauthIntrospectionURL.
+func NewOAuthTokenintrospectionAllKV(cfg *OpenIDConfig, timeout time.Duration) filters.Spec {
+	return newOAuthTokenintrospectionFilter(checkOAuthTokenintrospectionAllKV, cfg, timeout)
+}
+
+func NewOAuthTokenintrospectionAnyClaims(cfg *OpenIDConfig, timeout time.Duration) filters.Spec {
+	return newOAuthTokenintrospectionFilter(checkOAuthTokenintrospectionAnyClaims, cfg, timeout)
+}
+
+func NewOAuthTokenintrospectionAllClaims(cfg *OpenIDConfig, timeout time.Duration) filters.Spec {
+	return newOAuthTokenintrospectionFilter(checkOAuthTokenintrospectionAllClaims, cfg, timeout)
+}
+
+func newOAuthTokenintrospectionFilter(typ roleCheckType, cfg *OpenIDConfig, timeout time.Duration) filters.Spec {
+	return &tokenIntrospectionSpec{
+		typ:              typ,
+		issuerURL:        cfg.Issuer,
+		introspectionURL: cfg.IntrospectionEndpoint,
+		config:           cfg,
+		timeout:          timeout,
+	}
+}
+
+func GetOpenIDConfig(issuerURL string) (*OpenIDConfig, error) {
+	u, err := url.Parse(issuerURL + TokenIntrospectionConfigPath)
+	if err != nil {
+		return nil, err
+	}
+
+	var cfg OpenIDConfig
+	err = jsonGet(u, "", &cfg, http.DefaultClient)
+	return &cfg, err
+}
+
+func (s *tokenIntrospectionSpec) Name() string {
+	switch s.typ {
+	case checkOAuthTokenintrospectionAnyClaims:
+		return OAuthTokenintrospectionAnyClaimsName
+	case checkOAuthTokenintrospectionAllClaims:
+		return OAuthTokenintrospectionAllClaimsName
+	case checkOAuthTokenintrospectionAnyKV:
+		return OAuthTokenintrospectionAnyKVName
+	case checkOAuthTokenintrospectionAllKV:
+		return OAuthTokenintrospectionAllKVName
+	}
+	return AuthUnknown
+}
+
+func (s *tokenIntrospectionSpec) CreateFilter(args []interface{}) (filters.Filter, error) {
+	sargs, err := getStrings(args)
+	if err != nil {
+		return nil, err
+	}
+	if len(sargs) == 0 {
+		return nil, filters.ErrInvalidFilterParameters
+	}
+
+	ac, err := newAuthClient(s.introspectionURL, s.timeout)
```

Comment:
i think that the authClient should be created at the time when the filter spec is created, which means once on starting up skipper. If we create the authClient every time a new filter instance is created, there will be a new goroutine created in createHTTPClient and never be cleaned up.

Reply:
You are right. If I now think about, it might also makes sense to have multiple `issuer` allowed, so I could initialize atuhClient per issuer. The problem is how to have this accessible for the filter.

In the single issuer case, I can intialize the might failing part in skipper.go and put the client initialization into the spec. If we create a filter from the spec I would just copy the pointer.
In this way we would have for N filters of these type N times called close on the channel, which would panic.
Any idea how to resolve this easily?

Reply:
i would put it into skipper.go, and pointer can be passed from the spec to the filter instance.